flang/lib/Semantics/resolve-directives.cpp

```diff
@@ -2276,7 +2281,8 @@ void OmpAttributeVisitor::Post(const parser::Name &name) {
       if (Symbol * found{currScope().FindSymbol(name.source)}) {
         if (symbol != found) {
           name.symbol = found; // adjust the symbol within region
-        } else if (GetContext().defaultDSA == Symbol::Flag::OmpNone &&
+        } else if ((HaveOpenMPDefaultNone ||
```

Comment:
Would `HaveOpenMPDefaultNone` be set for those constructs where data-sharing is not allowed as well?

Reply:
As far as I understand the code, this is the place where all the handling of the implicit and explicit data-sharing checks are happening.  So, this should happen for all directives that allow for having data-sharing attribute clauses.

Reply:
I think this code used to be entered only if `defaultDSA` is set to `None`. And this setting is explicitly done in resolve-directives while going through the `OmpDefaultClause` in https://github.com/llvm/llvm-project/blob/4b56345895729fda3bc3c094bc3f237ba3a49686/flang/lib/Semantics/resolve-directives.cpp#L2019. Whereas the `HaveOpenMPDefaultNone` flag is set just based on the command line flag setting. We probably have to enable this flag only when datasharing clauses are supported.

To give an example where this will fire incorrectly.

```
subroutine sb1(arr, n)
  integer :: n
  integer :: arr(n)
  !$omp workshare
    arr = n
  !$omp end workshare
end subroutine
```

Reply:
This is exactly what this option is supposed to do.  If I misread your comment, please bear with me and explain.  For

```Fortran
subroutine sb1(arr, n)
  integer :: n
  integer :: arr(n)
  !$omp workshare
    arr = n
  !$omp end workshare
end subroutine
```

The `-fopenmp-default-none` acts as if the developer has written this:

```Fortran
subroutine sb1(arr, n)
  integer :: n
  integer :: arr(n)
  !$omp workshare default(none)
    arr = n
  !$omp end workshare
end subroutine
```

So, for both the error would be the same:

```
error: Semantic errors in /net/home/micha/projects/llvm/tmp/datasharing.f90
/net/home/micha/projects/llvm/tmp/datasharing.f90:16:5: error: The DEFAULT(NONE) clause requires that 'arr' must be listed in a data-sharing attribute clause
      arr = n
      ^^^
/net/home/micha/projects/llvm/tmp/datasharing.f90:16:11: error: The DEFAULT(NONE) clause requires that 'n' must be listed in a data-sharing attribute clause
      arr = n
```

Reply:
There is probably a bug here since resolve-directives happens before check-omp-structure.

Reply:
Oh boy...  Yes, I do not like these 20th century features. :-)  Thanks for catching this!